libsolutil/Dominator.h

```diff
@@ -0,0 +1,377 @@
+		while (idomIdx != 0)
+		{
+			solAssert(m_immediateDominators.at(idomIdx) < idomIdx);
+			dominators.emplace_back(m_verticesInDFSOrder.at(idomIdx));
+			idomIdx = m_immediateDominators[idomIdx];
+		}
+		// The loop above discovers the dominators in the reverse order
+		// i.e. from the given vertex upwards to the entry node (the root of the dominator tree).
+		// And the entry vertex always dominates all other vertices.
+		dominators.emplace_back(m_verticesInDFSOrder[0]);
+
+		return dominators;
+	}
+
+	/// Find all dominators of a node _v
+	/// @note for a vertex ``_v``, the _v’s inclusion in the set of dominators of ``_v`` is implicit.
+	std::vector<VId> dominatorsOf(V const& _v) const
+	{
+		return dominatorsOf(_v.id);
+	}
+
+private:
+	std::vector<DfsIndex> findDominators(V const& _entry)
+	{
+		// semi(w): The DFS index of the semidominator of ``w``.
+		std::vector<DfsIndex> semi{};
+		// parent(w): The index of the vertex which is the parent of ``w`` in the spanning
+		// tree generated by the DFS.
+		std::vector<DfsIndex> parent{};
+		// label(w): The index of the vertex ``w`` with the minimum semidominator in the path
+		// to its parent.
+		std::vector<DfsIndex> label{};
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<DfsIndex>> predecessors{};
+
+		// step 1
+		std::set<VId> visited;
+		// The number of vertices reached during the DFS.
+		// The vertices are indexed based on this number.
+		DfsIndex dfIdx = 0;
+		auto dfs = [&](V const& _v, auto _dfs) -> void {
+			auto [_, inserted] = visited.insert(_v.id);
+			if (!inserted)
+				return;
+			m_verticesInDFSOrder.emplace_back(_v.id);
+			m_dfsIndexByVertexId[_v.id] = dfIdx;
+			semi.emplace_back(dfIdx);
+			label.emplace_back(dfIdx);
+			// NOTE: Since each vertex is only visited once, incrementing dfIdx before visiting successors ensures
+			// it points to the next vertex to be visited (i.e., w).
+			// This is because we do not know the index of a vertex before visiting them.
+			dfIdx++;
+			ForEachSuccessor{}(_v, [&](V const& w) {
+				if (!visited.count(w.id))
+				{
+					// Here, ``dfIdx`` is the index of the vertex ``w`` in the DFS order.
+					if (dfIdx >= parent.size())
+						parent.resize(dfIdx + 1);
+					parent[dfIdx] = m_dfsIndexByVertexId[_v.id];
+					_dfs(w, _dfs);
+				}
+				if (m_dfsIndexByVertexId[w.id] >= predecessors.size())
+					predecessors.resize(m_dfsIndexByVertexId[w.id] + 1);
+				predecessors[m_dfsIndexByVertexId[w.id]].insert(m_dfsIndexByVertexId[_v.id]);
+			});
+		};
+		dfs(_entry, dfs);
+		solAssert(dfIdx == m_verticesInDFSOrder.size());
+		solAssert(dfIdx == visited.size());
+		solAssert(dfIdx == semi.size());
+		solAssert(dfIdx == label.size());
+
+		size_t numVertices = visited.size();
+		if (predecessors.size() < numVertices)
+			predecessors.resize(numVertices);
+		if (parent.size() < numVertices)
+			parent.resize(numVertices);
```

Comment:
By the way, the logic in this DFS traversal is still tripping me up every time I get back to this code. The naming, the way you have 3 nested functions here and each one reaches outside of its scope, the non-linear way in which `parent` and `predecessor` are filled out...

I tried simplifying it and I think that something like this would make it simpler to process:

```c++
		// step 1
		// The vertices are assigned indices in DFS order.
		std::set<VId> visited;
		DfsIndex nextUnusedDFSIndex = 0;

		auto dfs = [&](V const& _v, auto _dfs) -> void {
			auto [_, inserted] = visited.insert(_v.id);
			if (!inserted)
				return;

			m_verticesInDFSOrder.emplace_back(_v.id);
			m_dfsIndexByVertexId[_v.id] = nextUnusedDFSIndex;
			semi.emplace_back(nextUnusedDFSIndex);
			label.emplace_back(nextUnusedDFSIndex);
			nextUnusedDFSIndex++;

			ForEachSuccessor{}(_v, [&](V const& _successor) {
				if (visited.count(_successor.id) == 0)
				{
					parent.push_back(m_dfsIndexByVertexId[_v.id]);
					predecessors.push_back({m_dfsIndexByVertexId[_v.id]});
					_dfs(_successor, _dfs);
				}
				else
				{
					solAssert(m_dfsIndexByVertexId[_successor.id] < predecessors.size());
					predecessors[m_dfsIndexByVertexId[_successor.id]].insert(m_dfsIndexByVertexId[_v.id]);
				}
			});
		};

		parent.emplace_back(std::numeric_limits<DfsIndex>::max());
		predecessors.emplace_back();
		dfs(_entry, dfs);

		size_t numVertices = visited.size();
		solAssert(nextUnusedDFSIndex == numVertices);
		solAssert(m_verticesInDFSOrder.size() == numVertices);
		solAssert(visited.size() == numVertices);
		solAssert(semi.size() == numVertices);
		solAssert(label.size() == numVertices);
		solAssert(predecessors.size() == numVertices);
		solAssert(parent.size() == numVertices);
```

I'd maybe even go as far as to put this part in a separate function, since it's completely independent from the rest of the algorithm. The tables it fills out can be pretty much treated as constants after step 1 (except for `semi` and `label`, but these could also very easily be filled out separately).

Reply:
Yeah, it could be moved and pre-computed indeed. But we have done so many refactors, that I would rather postpone this one a bit, so we can finally merge a version of this PR and focus on the use of it, i.e. finish the Yul SSA transform.

Reply:
Sure, I also want to get done with it, which is why I went for doing fixups myself rather than just dropping more suggestions :) I included the above as one of them so the most important thing is done. I could not make up my mind whether it's better to have it in a separate function or not so if you don't want to make more changes here, it's fine to leave it as it is now.